test/test_nn.py

```diff
@@ -5526,8 +5526,8 @@ def compare_cpu_gpu(outputs_cpu, outputs_gpu):
                     self.assertEqual(cpu_weight.grad.data, gpu_weight.grad.data, atol=5e-5)
 
         for module in (nn.RNN, nn.LSTM, nn.GRU):
-            for bias, bidirectional, batch_first, contig, variable_len, lens_as_tensor \
-                    in product((True, False), repeat=6):
+            for bias, bidirectional, type_2, batch_first, contig, variable_len, lens_as_tensor \
```

Comment:
This test check CPU vs GPU and also covers ROCM, so all code is exercised. However it just checks for consistency between CPU and GPU implementations, not correctness (as far as I can tell). There are several other tests for `nn.RNN` & co with `bidirectional=True` that you could run for type1 and type2.

Reply:
test coverage looks much better now, thanks